### PR TITLE
Bugfix/374 employers purifier wrecking certain job ads

### PR DESCRIPTION
**Pull request by developerdino**
This is a major update to how this package cleans html input. Many filters have been updated to account for many more scenarios and outstanding bugs.

### CODE REVIEW

File: src/Filters/ConvertPsuedoHeadings.php [hunk 1]
```diff
@@ -33,7 +33,13 @@ public function apply(string $html): string
         );
 
         $html = preg_replace(
-            '/(<p>(?:\s+)?<strong>(.{0,80}?)<\/strong>(<br *\/?>)?(?:\s+)?<\/p>)+/si',
+            '/(<strong>)\s*?(<\/strong>)/im',
+            ' ',
+            str_ireplace('\n', ' ', $html)
```
Comment (nasimoyz) [inline src/Filters/ConvertPsuedoHeadings.php]: Should we replace line breaks in a different filter? Might fit more in `RemoveSpaces`

File: tests/Fixtures/LoadsFixtures.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+trait LoadsFixtures
+{
+    /**
+     * Loads fixtures for use in data providers.
+     *
+     * @param string $directory
+     * @param int $limit
+     * @return iterable
+     */
+    private function fixtureData(string $directory, int $limit): iterable
```
Comment (nasimoyz) [inline tests/Fixtures/LoadsFixtures.php]: Can we determine the `limit` from examining the directory?

File: tests/Functional/BeaumontConsultingTest.php [hunk 3]
```diff
@@ -3,55 +3,33 @@
...
     }
 
     /**
      * @test
+     * @dataProvider data
+     * @param $before
+     * @param $after
      */
-    public function it_correctly_purifies_fixture_3()
+    final public function it_correctly_purifies(string $before, string $after) :void
```
Comment (nasimoyz) [inline tests/Functional/BeaumontConsultingTest.php]: Space should come after colon, not before

Comment (nasimoyz) [inline tests/Functional/BeaumontConsultingTest.php]: Not sure if it's possible (if the test will still run), but could we add this method to the trait seeing as we re-use it in each test?

File: tests/Functional/BeaumontConsultingTest.php [hunk 4]
```diff
@@ -3,55 +3,33 @@
...
-        $this->assertEquals($purified, $after);
-    }
+    use LoadsFixtures;
 
     /**
-     * @test
+     * Data for it_correctly_purifies test
      */
-    public function it_correctly_purifies_fixture_2()
+    public final function data(): iterable
```
Comment (nasimoyz) [inline tests/Functional/BeaumontConsultingTest.php]: Do all these test methods need to be `final`? Is there a risk of the method being overridden?

**Review by nasimoyz — COMMENTED** [review]
It looks amazing! I just added some small questions

File: src/Filters/ConvertPsuedoHeadings.php (hunk 1 above)
Comment (developerdino) [reply src/Filters/ConvertPsuedoHeadings.php]: Yeah not sure we need this still, I think it is a remnant of something I was trying while in dev.

File: tests/Fixtures/LoadsFixtures.php (hunk 2 above)
Comment (developerdino) [reply tests/Fixtures/LoadsFixtures.php]: I did consider globbing the directory, will see if I can simplify.

File: tests/Functional/BeaumontConsultingTest.php (hunk 4 above)
Comment (developerdino) [reply tests/Functional/BeaumontConsultingTest.php]: It's my strict setting in my IDE, I can't find where it is :). Will turn this off and remove all of them as it is a bit full on.

File: tests/Functional/BeaumontConsultingTest.php (hunk 3 above)
Comment (developerdino) [reply tests/Functional/BeaumontConsultingTest.php]: Yes it will and done.

Comment (developerdino) [reply tests/Functional/BeaumontConsultingTest.php]: FIxed

File: tests/Fixtures/LoadsFixtures.php (hunk 2 above)
Comment (developerdino) [reply tests/Fixtures/LoadsFixtures.php]: Done

File: src/Filters/ConvertPsuedoHeadings.php (hunk 1 above)
Comment (developerdino) [reply src/Filters/ConvertPsuedoHeadings.php]: Removed and tests still passing.

File: tests/Fixtures/LoadsFixtures.php (hunk 2 above)
Comment (nasimoyz) [reply tests/Fixtures/LoadsFixtures.php]: Great solution!

**Review by nasimoyz — APPROVED** [review]
LGTM 👍

File: tests/Fixtures/LoadsFixtures.php [hunk 5]
```diff
@@ -0,0 +1,51 @@
+<?php
```
Comment (IgorNadj) [inline tests/Fixtures/LoadsFixtures.php]: Looks good but what is the value of changing this? It's gone from simple but readable data loading read to less simple and not so readable.

**Comment by IgorNadj** [comment]
1f29809dd3910fa97ea4c429dc7a4f4e2d20865d has formatting/cleanup changes alongside behaviour changes. Hard to read, split these into two separate commits.

Same with e06dd5dff96c7368acec0e8bcda032d20af160ac and 28b2639bc04d705d414b231d01a40b2d924800ec

**Comment by IgorNadj** [comment]
fc11b381241b33029581843fe4cb99051ab9abb1 looks like a change to behaviour without an associated test case change. Is not not covered by tests or is the test change in another commit?

**Comment by IgorNadj** [comment]
084a78b60db9f2ce4ce596c0a82cf7168c155b1e undos changes that you made in the same PR. Please simplify the PR by rewriting history to have never changed them in the first place, this makes it much easier for a reviewer to follow :) 

This is done with `git rebase -i HEAD~10`, change the commits you want to edit to `edit`, then change those commits to have never made those changes, then `git rebase --continue`.

File: tests/Fixtures/LoadsFixtures.php (hunk 5 above)
Comment (developerdino) [reply tests/Fixtures/LoadsFixtures.php]: Way easier to implement new tests and follows a standard paradigm.

**Comment by developerdino** [comment]
The whole commit thing I need to get used to. Never had to rewrite history before as you are the first colleague that reviews in this way. I will try to do more clean up on branches going forward.

File: tests/Fixtures/LoadsFixtures.php (hunk 5 above)
Comment (IgorNadj) [reply tests/Fixtures/LoadsFixtures.php]: Sorted

**Comment by IgorNadj** [comment]
Have discussed, will assume changes are good. Happy to merge.

**Pull request merged**